sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -769,6 +770,63 @@ object ReorderJoin extends Rule[LogicalPlan] with PredicateHelper {
 }
 
 /**
+ * Conversion of outer joins, if the local predicates can restrict the result sets so that
+ * all null-supplying rows are eliminated
+ *
+ * - full outer -> inner if both sides have such local predicates
+ * - left outer -> inner if the right side has such local predicates
+ * - right outer -> inner if the left side has such local predicates
+ * - full outer -> left outer if only the left side has such local predicates
+ * - full outer -> right outer if only the right side has such local predicates
+ *
+ * This rule should be executed before pushing down the Filter
+ */
+object OuterJoinConversion extends Rule[LogicalPlan] with PredicateHelper {
```

Comment:
You are right. `OuterJoinElimination` might sound better. Let me rename it today. 

Since `full outer` is `union distinct` of `left outer` and `right outer`, we are removing `right outer` from `full outer` in conversion from `full outer` to `left outer`.